tests/std/tests/P0896R4_views_take_while/test.cpp

```diff
@@ -0,0 +1,493 @@
+        const same_as<iterator_t<R>> auto i = r.begin();
+        if (!is_empty) {
+            assert(*i == *begin(expected));
+            assert(*r.begin() == *begin(expected));
+            assert(*r.begin() == *begin(expected));
+        }
+
+        if constexpr (copyable<V>) {
+            auto r2                              = r;
+            const same_as<iterator_t<R>> auto i2 = r2.begin();
+            if (!is_empty) {
+                assert(*i2 == *i);
+                assert(*r2.begin() == *i2);
+                assert(*r2.begin() == *i2);
+            }
+        }
+
+        const same_as<iterator_t<const R>> auto ic = as_const(r).begin();
+        if (!is_empty) {
+            assert(*ic == *begin(expected));
+            assert(*as_const(r).begin() == *begin(expected));
+            assert(*as_const(r).begin() == *begin(expected));
+        }
+
+        if constexpr (copyable<V>) {
+            auto rc2                                    = as_const(r);
+            const same_as<iterator_t<const R>> auto ic2 = rc2.begin();
+            if (!is_empty) {
+                assert(*ic2 == *ic);
+                assert(*rc2.begin() == *ic2);
+                assert(*rc2.begin() == *ic2);
+            }
+        }
+    }
+
+    // Validate take_while_view::end
+    STATIC_ASSERT(CanMemberEnd<R>);
+    STATIC_ASSERT(CanMemberEnd<const R>);
+    STATIC_ASSERT(CanEnd<const R&> == ranges::range<const V>);
+    STATIC_ASSERT(!common_range<R>);
+    STATIC_ASSERT(!common_range<const R>);
+    if (!is_empty) {
+        [[maybe_unused]] same_as<ranges::sentinel_t<R>> auto s        = r.end();
+        [[maybe_unused]] same_as<ranges::sentinel_t<const R>> auto sc = as_const(r).end();
+
+        if (forward_range<V>) { // intentionally not if constexpr
+            // Compare with const / non const iterators
+            assert(s != r.begin());
+            assert(s != as_const(r).begin());
+            assert(sc != r.begin());
+            assert(sc != as_const(r).begin());
+
+            // Compare with end of range
+            if constexpr (common_range<V>) {
+                assert(s == s.base());
+                assert(s == sc.base());
+                assert(sc == s.base());
+                assert(sc == sc.base());
+            } else {
+                if constexpr (forward_range<V>) {
+                    const auto length = 8; // NB: depends on the test data
+                    assert(s == next(r.begin(), length));
+                    assert(s == next(as_const(r).begin(), length));
+                    assert(sc == next(r.begin(), length));
+                    assert(sc == next(as_const(r).begin(), length));
+                }
+            }
+
+            // Compare with iterator whose predicate evaluates to false
+            if constexpr (forward_range<V>) {
+                const auto length = 4; // NB: depends on the test data
+                assert(s == next(r.begin(), length));

```

Comment:
What? How do these succeed after 260-263 succeed?

Reply:
This should be the second thing:
```cpp
            // clang-format off
            template <bool _OtherConst = !_Const>
                requires sentinel_for<_Base_sentinel, _Maybe_const_iter<_OtherConst>> // Per Resolution of LWG-3449
            friend constexpr bool operator==(const _Maybe_const_iter<_OtherConst>& _Left, const _Sentinel& _Right) {
                // clang-format on
                return _Right._Last == _Left || !_STD invoke(*_Right._Pred, *_Left);
            }
```

Reply:
Ok, I see what's going on. Here we validate the end of the range, and 262-266 is formally UB due to advancing iterators past said end of the range. Taking advantage of our knowledge of the implementation is convenient here since it lets us test both branches of the iterator/sentinel equality with a single set of test data.

This feels icky to me. For 262-266 could we instead create a different view over the same elements with a permissive predicate, to avoid relying on the UB?